classes/profile.php

```diff
@@ -193,12 +193,32 @@ public function save_record(): int {
         $this->raw_set('timemodified', $now);
         $this->raw_set('usermodified', $USER->id);
 
+        $storeduser = (int) $this->raw_get('userid');
+
         // We may not have obtained a valid userid when the profile record was created.
-        // If the current profile userid is 0, but there's a valid $USER->id, update it.
-        if ($USER->id && ! $this->raw_get('userid')) {
+        // If the stored userid is 0, and there's now a valid $USER->id, update the stored userid.
+        if ($USER->id && !$storeduser) {
+            $this->raw_set('userid', $USER->id);
+        }
+
+        // If current userid isn't guest, but stored user is guest, update it.
+        $guestrole = get_guest_role();
+        $guestid = (int) $guestrole->id;
+        if ($USER->id !== $guestid && $storeduser === $guestid) {
             $this->raw_set('userid', $USER->id);
         }
```

Comment:
I feel like for these 3 conditional blocks, the outcome is the same (if any were true), i.e. it will update the userid on the profile.

I see potential overlap in checks, e.g. if you pass the first check, you're always checking for guest role, and if you were in CLI, you're always checking admin users. 

It might be better as one big if condition or calculated in a separate method/function and set if required. Personally I'm leaning towards handling this in a separate private method, maybe something like `get_resolved_user` which returns the userid and updates it if its different.